src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,453 @@
+        if (line.isClosed())
+        {
+            flag = createFlag(object,
+                    this.getLocalizedInstruction(FALLBACK_INSTRUCTIONS.indexOf(CIRCULAR_WATERWAY),
+                            object.getOsmIdentifier()),
+                    Collections.singletonList(line.asPolyLine().first()));
+        }
+        else if (uphill && this.minResolutionDistance
+                .isGreaterThanOrEqualTo(this.elevationUtils.getResolution(first)))
+        {
+            flag = createUphillFlag(object, first);
+        }
+        else if (isValidEndToCheck(line.getAtlas(), last) && !doesWaterwayEndInSink(line))
+        {
+            if (uphill)
+            {
+                flag = createUphillFlag(object, first);
+            }
+            else if (!endsWithBoundaryNode(object))
+            {
+                flag = createFlag(object,
+                        this.getLocalizedInstruction(
+                                FALLBACK_INSTRUCTIONS.indexOf(DOES_NOT_END_IN_SINK),
+                                object.getOsmIdentifier()),
+                        Collections.singletonList(last));
+            }
+        }
+        final LineItem crossed = intersectsAnotherWaterWay(line);
+        if (flag == null && crossed != null)
+        {
+            final Iterator<Location> intersections = crossed.asPolyLine()
+                    .intersections(line.asPolyLine()).iterator();
+            if (intersections.hasNext())
+            {
+                flag = createFlag(Sets.hashSet(object, crossed),
+                        this.getLocalizedInstruction(
+                                FALLBACK_INSTRUCTIONS.indexOf(CROSSES_WATERWAY),
+                                object.getOsmIdentifier(), crossed.getOsmIdentifier()),
+                        Arrays.asList(intersections.next()));
+            }
+        }
+        if (flag == null)
+        {
+            return Optional.empty();
+        }
+        super.markAsFlagged(object.getOsmIdentifier());
+        return Optional.of(flag);
+    }
+
+    @Override
+    protected List<String> getFallbackInstructions()
+    {
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    private CheckFlag createUphillFlag(final AtlasObject object, final Location first)
+    {
+        return createFlag(object,
+                this.getLocalizedInstruction(FALLBACK_INSTRUCTIONS.indexOf(GOES_UPHILL),
+                        object.getOsmIdentifier(),
+                        this.elevationUtils.getResolution(first).asMeters()));
+    }
+
+    private boolean doesWaterwayEndInSink(final LineItem line)
+    {
+        return doesLineEndOnWaterway(line) || doesLineEndInSink(line) || doesLineEndInOcean(line);
+    }
+
+    private LineItem intersectsAnotherWaterWay(final LineItem line)
+    {
+        final Atlas atlas = line.getAtlas();
+        final Iterable<LineItem> intersectinWaterways = atlas.lineItemsIntersecting(line.bounds(),
```

Comment:
nit: intersectingWaterways

Reply:
pick: nit (I'll fix it, it would bug me as well if I didn't...)